### PR TITLE
Falcon40b prefill bringup

### PR DESCRIPTION
**Pull request by johanna-rock-tt**
Adds support for falcon40b prefill for variable sequences lengths up to S=2048.
- Uses separate fwd methods for sharded decode (unchanged) and prefill
- Prefill is mostly DRAM interleaved, exception is layernorm which is block sharded for functional reasons

**PCC**
1 layer S=128 and S=2048: 0.99 / 0.99 / 0.99
60 layer S=128: 0.92 / 0.98 / 0.94

**Performance**
60 layer S=128: 1173 tok/s
60 layer S=2048: 1961 tok/s

### CODE REVIEW

**Review by boris-drazic — APPROVED** [review]

**Review by TT-BrianLiu — CHANGES_REQUESTED** [review]
Can we try to merge with the existing falcon40b code? Also, squash the commits as much as possible.

**Comment by johanna-rock-tt** [comment]
I applied my changes to the existing models/demos/falcon40b code. Still need to run all tests.

File: models/demos/falcon40b/demo/demo.py [hunk 1]
```diff
@@ -195,8 +195,8 @@ def run_falcon_demo_kv(
     prefill_on_host,
 ):
     torch.manual_seed(0)
-    for device in devices:
-        device.enable_program_cache()
+    # for device in devices:
+    #     device.enable_program_cache()
```
Comment (TT-BrianLiu) [inline models/demos/falcon40b/demo/demo.py:199]: Do we have to turn off program cache? Can we enable and leave what is working on `main`?

File: models/demos/falcon40b/demo/demo.py [hunk 2]
```diff
@@ -424,8 +424,8 @@ def run_falcon_demo_kv(
 
     print_output_prompts(generated_ids, tokenizer)
 
-    for device in devices:
-        device.disable_and_clear_program_cache()
+    # for device in devices:
+    #     device.disable_and_clear_program_cache()
```
Comment (TT-BrianLiu) [inline models/demos/falcon40b/demo/demo.py:428]: Do we have to turn off program cache?

File: models/demos/falcon40b/demo/demo.py [hunk 3]
```diff
@@ -591,7 +591,7 @@ def test_demo(
     model_location_generator,
     get_tt_cache_path,
     all_devices,
-    use_program_cache,
+    # use_program_cache, # TODO: remove workaround for avoiding PCC issues when fixed
```
Comment (TT-BrianLiu) [inline models/demos/falcon40b/demo/demo.py:594]: Do we have to turn off program cache? Also, if we have this, we don't need the manual enable in the demo code.

File: models/demos/falcon40b/tests/test_falcon_attention.py [hunk 4]
```diff
@@ -354,10 +363,12 @@ def test_FalconAttention_inference(
     model_location_generator,
     get_tt_cache_path,
     all_devices,
-    use_program_cache,
+    # use_program_cache, # TODO: remove workaround for avoiding PCC issues when fixed
```
Comment (TT-BrianLiu) [inline models/demos/falcon40b/tests/test_falcon_attention.py:366]: Don't remove.

File: models/demos/falcon40b/tests/test_falcon_causallm.py [hunk 5]
```diff
@@ -346,14 +352,12 @@ def test_FalconCausalLM_inference(
     model_location_generator,
     get_tt_cache_path,
     all_devices,
-    use_program_cache,
+    # use_program_cache, # TODO: remove workaround for avoiding PCC issues when fixed
 ):
```
Comment (TT-BrianLiu) [inline models/demos/falcon40b/tests/test_falcon_causallm.py:356]: Don't remove.

File: models/demos/falcon40b/tests/test_falcon_decoder.py [hunk 6]
```diff
@@ -373,10 +386,12 @@ def test_FalconDecoder_inference(
     model_location_generator,
     get_tt_cache_path,
     all_devices,
-    use_program_cache,
+    # use_program_cache, # TODO: remove workaround for avoiding PCC issues when fixed
```
Comment (TT-BrianLiu) [inline models/demos/falcon40b/tests/test_falcon_decoder.py:389]: Don't remove

File: models/demos/falcon40b/tests/test_falcon_end_to_end.py [hunk 7]
```diff
@@ -430,21 +445,24 @@ def test_FalconCausalLM_end_to_end_with_program_cache(
     model_location_generator,
     get_tt_cache_path,
     all_devices,
-    use_program_cache,
+    # use_program_cache, # TODO: remove workaround for avoiding PCC issues when fixed
 ):
```
Comment (TT-BrianLiu) [inline models/demos/falcon40b/tests/test_falcon_end_to_end.py:449]: Don't remove

File: models/demos/falcon40b/tests/test_falcon_mlp.py [hunk 8]
```diff
@@ -123,8 +139,13 @@ def test_FalconMLP_inference(
     model_location_generator,
     get_tt_cache_path,
     all_devices,
-    use_program_cache,
+    # use_program_cache, # TODO: remove workaround for avoiding PCC issues when fixed,
 ):
```
Comment (TT-BrianLiu) [inline models/demos/falcon40b/tests/test_falcon_mlp.py:143]: Don't remove

File: models/demos/falcon40b/tests/test_falcon_model.py [hunk 9]
```diff
@@ -341,14 +349,12 @@ def test_FalconModel_inference(
     model_location_generator,
     get_tt_cache_path,
     all_devices,
-    use_program_cache,
+    # use_program_cache, # TODO: remove workaround for avoiding PCC issues when fixed
 ):
```
Comment (TT-BrianLiu) [inline models/demos/falcon40b/tests/test_falcon_model.py:353]: Don't remove

File: models/demos/falcon40b/tests/test_perf_falcon.py [hunk 10]
```diff
@@ -460,8 +459,13 @@ def test_perf_bare_metal(
     model_location_generator,
     get_tt_cache_path,
     all_devices,
-    use_program_cache,
```
Comment (TT-BrianLiu) [inline models/demos/falcon40b/tests/test_perf_falcon.py:463]: Don't remove

**Review by TT-BrianLiu — CHANGES_REQUESTED** [review]
We should leave program caching in the tests. CI tests currently use program_cache. If we remove, we are testing less in CI

**Comment by johanna-rock-tt** [comment]
> We should leave program caching in the tests. CI tests currently use program_cache. If we remove, we are testing less in CI

Program cache enabled gives very low PCC results and also garbage outputs for the demo. I think it's not an issue for decode but it is for prefill. For longer sequence lengths (S=2048) this results in a PCC of 0.49 for one layer, for shorter ones it's also lower PCC but not that dramatic (0.8-0.9) but even for S=32 the demo output is garbage with program cache enabled.

See [this issue.](https://github.com/tenstorrent-metal/tt-metal/issues/7159)

Suggestion: I can add another test where program cache is enabled to run the decode tests in CI with program cache but I'd keep the tests where it is temporarily disabled - otherwise we can't add prefill to CI at all and also everybody trying out the demo and/or the tests is going to see bad outputs.

**Review by TT-BrianLiu — APPROVED** [review]

**Review by tt-rkim — APPROVED** [review]

**Review by cfjchu — APPROVED** [review]

**Pull request merged**